src/Sylius/Component/Core/OrderProcessing/PaymentProcessor.php

```diff
@@ -51,11 +51,22 @@ public function processOrderPayments(OrderInterface $order)
      */
     private function setPaymentMethodIfNeeded(OrderInterface $order, PaymentInterface $payment)
     {
-        $lastCancelledPayment = $order->getLastPayment(PaymentInterface::STATE_CANCELLED);
         $lastNewPayment = $order->getLastPayment(PaymentInterface::STATE_NEW);
 
-        if (!$lastNewPayment && $lastCancelledPayment) {
-            $payment->setMethod($lastCancelledPayment->getMethod());
+        if ($lastNewPayment) {
+            return;
         }
+
+        $lastCancelledOrFailedPayment = $order->getLastPayment(PaymentInterface::STATE_CANCELLED);
+
+        if (!$lastCancelledOrFailedPayment) {
+            $lastCancelledOrFailedPayment = $order->getLastPayment(PaymentInterface::STATE_FAILED);
+        }
```

Comment:
How about extracting four lines above into separate method? Then we could use returned variable in following if. Moreover, variable should be named ``$lastPayment`` or sth like this, I suppose, as ``$lastCancelledOrFailedPayment`` smells bad :)